hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/ReplicatedMapEventPublishingService.java

```diff
@@ -187,13 +189,26 @@ public void fireEntryListenerEvent(Data key, Data oldValue, Data value, EntryEve
         }
         EntryEventData eventData = new EntryEventData(name, name, caller, key, value, oldValue, eventType.getType());
         for (EventRegistration registration : registrations) {
-            EventFilter filter = registration.getFilter();
-            boolean publish = filter == null || filter.eval(key);
-            if (publish) {
-                eventService.publishEvent(SERVICE_NAME, registration, eventData, key.hashCode());
+            if (!shouldPublish(key, oldValue, value, eventType, registration.getFilter())) {
+                continue;
             }
+            eventService.publishEvent(SERVICE_NAME, registration, eventData, key.hashCode());
         }
     }
 
+    private boolean shouldPublish(Data key, Data oldValue, Data value, EntryEventType eventType, EventFilter filter) {
+        QueryEntry queryEntry = null;
+        if (filter instanceof ReplicatedQueryEventFilter) {
```

Comment:
Can it be that a different type of filter than a ReplicatedQueryEventFilter is going to be passed? If so, is this not erronous behavior? It could very well be that I'm wrong; I'm trying to figure out if we should fail fast or not.. or if this is a normal flow.